File: packages/sdk/src/resources/edges.ts
/**
 * Edges Resource
 *
 * Resource-based SDK interface for knowledge crystal edge management.
 * Supports all relationship types including 'contains' for hierarchy (ADR-055).
 */

import { unwrapData, requireArray } from "../validate.js";
import { BaseResource } from "./base.js";
import type {
  KnowledgeCrystalEdge,
  CreateKnowledgeCrystalEdgeParams,
  UpdateKnowledgeCrystalEdgeParams,
  ListKnowledgeCrystalEdgesParams,
} from "../types/knowledge-crystal-edge.js";

const RESOURCE = "edges";

// ============================================================================
// API Response Types
// ============================================================================

interface ApiSuccessResponse<T> {
  data: T;
  meta?: {
    pagination?: {
      total: number;
      limit: number;
      offset?: number;
      hasMore: boolean;
    };
  };
}

// ============================================================================
// Edges Resource
// ============================================================================

/**
 * Edges Resource - manages edges between knowledge crystal nodes.
 * Supports all relationship types including 'contains' for hierarchy.
 */
export class EdgesResource extends BaseResource {
  /**
   * Create an edge between two knowledge crystal nodes
   */
  async create(params: CreateKnowledgeCrystalEdgeParams): Promise<KnowledgeCrystalEdge> {
    const response = await this.request<ApiSuccessResponse<KnowledgeCrystalEdge>>(
      "POST",
      "/v1/edges",
      params
    );
    return unwrapData<KnowledgeCrystalEdge>(response, "POST /v1/edges", RESOURCE);
  }

  /**
   * Get an edge by ID
   */
  async get(id: string): Promise<KnowledgeCrystalEdge> {
    const response = await this.request<ApiSuccessResponse<KnowledgeCrystalEdge>>(
      "GET",
      `/v1/edges/${encodeURIComponent(id)}`
    );
    return unwrapData<KnowledgeCrystalEdge>(response, "GET /v1/edges/{id}", RESOURCE);
  }

  /**
   * List edges with optional filters
   */
  async list(params?: ListKnowledgeCrystalEdgesParams): Promise<{
    edges: KnowledgeCrystalEdge[];
    total: number;
    hasMore: boolean;
  }> {
    const query = new URLSearchParams();

    if (params?.sourceId) {
      query.set("sourceId", params.sourceId);
    }
    if (params?.targetId) {
      query.set("targetId", params.targetId);
    }
    if (params?.relationship) {
      query.set("relationship", params.relationship);
    }
    if (params?.limit) {
      query.set("limit", String(params.limit));
    }
    if (params?.offset) {
      query.set("offset", String(params.offset));
    }

    const queryString = query.toString();
    const path = `/v1/edges${queryString ? `?${queryString}` : ""}`;

    const response = await this.request<ApiSuccessResponse<KnowledgeCrystalEdge[]>>(
      "GET",
      path
    );

    const edges = requireArray<KnowledgeCrystalEdge>(
      unwrapData<KnowledgeCrystalEdge[]>(response, "GET /v1/edges", RESOURCE),
      "GET /v1/edges",
      RESOURCE,
    );
    return {
      edges,
      total: response.meta?.pagination?.total ?? edges.length,
      hasMore: response.meta?.pagination?.hasMore ?? false,
    };
  }

  /**
   * Update an edge's metadata
   */
  async update(id: string, params: UpdateKnowledgeCrystalEdgeParams): Promise<KnowledgeCrystalEdge> {
    const response = await this.request<ApiSuccessResponse<KnowledgeCrystalEdge>>(
      "PATCH",
      `/v1/edges/${encodeURIComponent(id)}`,
      params
    );
    return unwrapData<KnowledgeCrystalEdge>(response, "PATCH /v1/edges/{id}", RESOURCE);
  }

  /**
   * Delete an edge
   */
  async delete(id: string): Promise<void> {
    await this.request<void>(
      "DELETE",
      `/v1/edges/${encodeURIComponent(id)}`
    );
  }
}

File: packages/sdk-python/engram/resources/edges.py
"""Edge resources for the Engram SDK."""
from __future__ import annotations

from typing import Optional, TYPE_CHECKING
from urllib.parse import quote

from engram._base import BaseResource, SyncBaseResource
from engram.types.common import PaginatedResult
from engram.types.knowledge_crystal import (
    KnowledgeCrystalEdge,
    CreateKnowledgeCrystalEdgeParams,
    UpdateKnowledgeCrystalEdgeParams,
    ListKnowledgeCrystalEdgesParams,
)

if TYPE_CHECKING:
    from engram.client import AsyncEngramClient, EngramClient


class EdgesResource(BaseResource):
    """Async resource for knowledge edges."""

    async def create(
        self,
        params: CreateKnowledgeCrystalEdgeParams,
    ) -> KnowledgeCrystalEdge:
        """Create an edge between knowledge items."""
        body = params.model_dump(by_alias=True, exclude_none=True)
        response = await self._request("POST", "/v1/edges", body)
        return KnowledgeCrystalEdge.model_validate(response["data"])

    async def get(self, id: str) -> KnowledgeCrystalEdge:
        """Get an edge by ID."""
        response = await self._request(
            "GET", f"/v1/edges/{quote(id, safe='')}"
        )
        return KnowledgeCrystalEdge.model_validate(response["data"])

    async def list(
        self,
        params: Optional[ListKnowledgeCrystalEdgesParams] = None,
    ) -> PaginatedResult[KnowledgeCrystalEdge]:
        """List edges."""
        qs: dict[str, str] = {}
        if params:
            if params.source_id:
                qs["sourceId"] = params.source_id
            if params.target_id:
                qs["targetId"] = params.target_id
            if params.relationship:
                qs["relationship"] = params.relationship
            if params.limit is not None:
                qs["limit"] = str(params.limit)
            if params.offset is not None:
                qs["offset"] = str(params.offset)
        response = await self._request("GET", "/v1/edges", params=qs if qs else None)
        return self._parse_list(response, KnowledgeCrystalEdge)

    async def update(
        self,
        id: str,
        params: UpdateKnowledgeCrystalEdgeParams,
    ) -> KnowledgeCrystalEdge:
        """Update an edge."""
        body = params.model_dump(by_alias=True, exclude_none=True)
        response = await self._request(
            "PATCH", f"/v1/edges/{quote(id, safe='')}", body
        )
        return KnowledgeCrystalEdge.model_validate(response["data"])

    async def delete(self, id: str) -> None:
        """Delete an edge."""
        await self._request("DELETE", f"/v1/edges/{quote(id, safe='')}")


class SyncEdgesResource(SyncBaseResource):
    """Sync resource for knowledge edges."""

    def create(
        self,
        params: CreateKnowledgeCrystalEdgeParams,
    ) -> KnowledgeCrystalEdge:
        """Create an edge between knowledge items."""
        body = params.model_dump(by_alias=True, exclude_none=True)
        response = self._request("POST", "/v1/edges", body)
        return KnowledgeCrystalEdge.model_validate(response["data"])

    def get(self, id: str) -> KnowledgeCrystalEdge:
        """Get an edge by ID."""
        response = self._request(
            "GET", f"/v1/edges/{quote(id, safe='')}"
        )
        return KnowledgeCrystalEdge.model_validate(response["data"])

    def list(
        self,
        params: Optional[ListKnowledgeCrystalEdgesParams] = None,
    ) -> PaginatedResult[KnowledgeCrystalEdge]:
        """List edges."""
        qs: dict[str, str] = {}
        if params:
            if params.source_id:
                qs["sourceId"] = params.source_id
            if params.target_id:
                qs["targetId"] = params.target_id
            if params.relationship:
                qs["relationship"] = params.relationship
            if params.limit is not None:
                qs["limit"] = str(params.limit)
            if params.offset is not None:
                qs["offset"] = str(params.offset)
        response = self._request("GET", "/v1/edges", params=qs if qs else None)
        return self._parse_list(response, KnowledgeCrystalEdge)

    def update(
        self,
        id: str,
        params: UpdateKnowledgeCrystalEdgeParams,
    ) -> KnowledgeCrystalEdge:
        """Update an edge."""
        body = params.model_dump(by_alias=True, exclude_none=True)
        response = self._request(
            "PATCH", f"/v1/edges/{quote(id, safe='')}", body
        )
        return KnowledgeCrystalEdge.model_validate(response["data"])

    def delete(self, id: str) -> None:
        """Delete an edge."""
        self._request("DELETE", f"/v1/edges/{quote(id, safe='')}")
